cmd/k8s-operator/tsrecorder_specs.go

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+//go:build !plan9
+
+package main
+
+import (
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	tsapi "tailscale.com/k8s-operator/apis/v1alpha1"
+	"tailscale.com/types/ptr"
+)
+
+func tsrStatefulSet(tsr *tsapi.TSRecorder, namespace string) *appsv1.StatefulSet {
+	return &appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            tsr.Name,
+			Namespace:       namespace,
+			Labels:          labels("tsrecorder", tsr.Name),
+			OwnerReferences: tsrOwnerReference(tsr),
+		},
+		Spec: appsv1.StatefulSetSpec{
+			Replicas: ptr.To[int32](1),
+			Selector: &metav1.LabelSelector{
+				MatchLabels: labels("tsrecorder", tsr.Name),
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      tsr.Name,
+					Namespace: namespace,
+					Labels:    labels("tsrecorder", tsr.Name),
+				},
+				Spec: corev1.PodSpec{
+					ServiceAccountName: tsr.Name,
+					Containers: []corev1.Container{
+						{
+							Name: "tsrecorder",
+							Image: func() string {
+								repo, tag := tsr.Spec.Image.Repo, tsr.Spec.Image.Tag
+								if repo == "" {
+									repo = "tailscale/tsrecorder"
+								}
+								if tag == "" {
+									tag = "stable"
+								}
+								return fmt.Sprintf("%s:%s", repo, tag)
+							}(),
+							Env: []corev1.EnvVar{
+								{
+									Name: "TS_AUTHKEY",
+									ValueFrom: &corev1.EnvVarSource{
+										SecretKeyRef: &corev1.SecretKeySelector{
+											LocalObjectReference: corev1.LocalObjectReference{
+												Name: tsr.Name,
+											},
+											Key: "authkey",
+										},
+									},
+								},
+								{
+									Name: "POD_NAME",
+									ValueFrom: &corev1.EnvVarSource{
+										FieldRef: &corev1.ObjectFieldSelector{
+											// Secret is named after the pod.
+											FieldPath: "metadata.name",
+										},
+									},
+								},
+								{
+									Name:  "TS_STATE",
+									Value: "kube:$(POD_NAME)",
+								},
+								{
+									Name:  "TSRECORDER_HOSTNAME",
+									Value: "$(POD_NAME)",
```

Comment:
@tomhjp wdyt if we make the hostname the TSRecorder resource name?
Right now we restrict to a single replica so this is fine and probably more what ppl want that <tsrecorder-resource-name>-0
When we have multiple replicas, they will anyway want them load balanced, rather than having to have <tsrecorder-resource-name>-0, <tsrecorder-resource-name>-1 etc - so we can put off the multiple replicas till we can do the internal load balancing?

Reply:
Agree 👍🏼 I was focusing on the `.status.url` value, but yes, we can change that once we have the load balancing